crypto/crmf/crmf_pbm.c

```diff
@@ -0,0 +1,222 @@
+ * returns 1 at success, 0 at error
+ */
+int OSSL_CRMF_pbm_new(const OSSL_CRMF_PBMPARAMETER *pbmp,
+                      const unsigned char *msg, size_t msglen,
+                      const unsigned char *sec, size_t seclen,
+                      unsigned char **mac, unsigned int *maclen)
+{
+    int mac_nid, hmac_md_nid = NID_undef;
+    const EVP_MD *m = NULL;
+    EVP_MD_CTX *ctx = NULL;
+    unsigned char basekey[EVP_MAX_MD_SIZE];
+    unsigned int bklen;
+    int64_t iterations;
+    int error = CRMF_R_CRMFERROR;
+
+    if (mac == NULL || pbmp == NULL || pbmp->mac == NULL ||
+            pbmp->mac->algorithm == NULL || msg == NULL || sec == NULL) {
+        error = CRMF_R_NULL_ARGUMENT;
+        goto err;
+    }
+    OPENSSL_free(*mac);
+    if ((*mac = OPENSSL_malloc(EVP_MAX_MD_SIZE)) == NULL) {
+        error = ERR_R_MALLOC_FAILURE;
+        goto err;
+    }
+
+    /*
+     * owf identifies the hash algorithm and associated parameters used to
+     * compute the key used in the MAC process.  All implementations MUST
+     * support SHA-1.
+     */
+    if ((m = EVP_get_digestbyobj(pbmp->owf->algorithm)) == NULL) {
+        error = CRMF_R_UNSUPPORTED_ALGORITHM;
+        goto err;
+    }
+
+    if ((ctx = EVP_MD_CTX_create()) == NULL) {
+        error = ERR_R_MALLOC_FAILURE;
+        goto err;
+    }
+
+    /* compute the basekey of the salted secret */
+    if (!(EVP_DigestInit_ex(ctx, m, NULL)))
+        goto err;
+    /* first the secret */
+    if (!EVP_DigestUpdate(ctx, sec, seclen))
+        goto err;
+    /* then the salt */
+    if (!EVP_DigestUpdate(ctx, pbmp->salt->data, pbmp->salt->length))
+        goto err;
+    if (!(EVP_DigestFinal_ex(ctx, basekey, &bklen)))
+        goto err;
+    if (!ASN1_INTEGER_get_int64(&iterations, pbmp->iterationCount)
+            || iterations < 100 /* min from RFC */
+            || iterations > OSSL_CRMF_PBM_MAX_ITERATION_COUNT) {
+        error = CRMF_R_BAD_PBM_ITERATIONCOUNT;
+        goto err;
+    }
+
+    /* the first iteration was already done above */
+    while (--iterations > 0) {
+        if (!(EVP_DigestInit_ex(ctx, m, NULL)))
+            goto err;
+        if (!EVP_DigestUpdate(ctx, basekey, bklen))
+            goto err;
+        if (!(EVP_DigestFinal_ex(ctx, basekey, &bklen)))
+            goto err;
+    }
+
+    /*
+     * mac identifies the algorithm and associated parameters of the MAC
+     * function to be used.  All implementations MUST support HMAC-SHA1
+     * [HMAC].      All implementations SHOULD support DES-MAC and Triple-
+     * DES-MAC [PKCS11].
+     */
+    mac_nid = OBJ_obj2nid(pbmp->mac->algorithm);
+
+    if (!EVP_PBE_find(EVP_PBE_TYPE_PRF, mac_nid, NULL, &hmac_md_nid, NULL) ||
+            ((m = EVP_get_digestbynid(hmac_md_nid)) == NULL)) {
+        error = CRMF_R_UNSUPPORTED_ALGORITHM;
+        goto err;
+    }
+    HMAC(m, basekey, bklen, msg, msglen, *mac, maclen);
+
+    /* cleanup */
+    OPENSSL_cleanse(basekey, bklen);
+    EVP_MD_CTX_destroy(ctx);
```

Comment:
ctx not cleaned up in error handler

Reply:
Fixed, also cleansing of basekey containing some secret data